File: src/main/java/calculadora/memento/Calculator.java
package calculadora.memento;

/**
 *
 * @author vedz11
 */
public class Calculator {
    private double currentValue = 0.0;

    public void add(double value) {
        currentValue += value;
    }

    public void subtract(double value) {
        currentValue -= value;
    }

    public void multiply(double value) {
        currentValue *= value;
    }

    public void divide(double value) {
        if (value != 0) {
            currentValue /= value;
        } else {
            throw new ArithmeticException("No se puede dividir por cero.");
        }
    }

    public void clear() {
        currentValue = 0.0;
    }

    public double getValue() {
        return currentValue;
    }

    public CalculatorMemento save() {
        return new CalculatorMemento(currentValue);
    }

    public void restore(CalculatorMemento memento) {
        this.currentValue = memento.getState();
    }
}

File: src/memento/Calculator.ts
import { CalculatorMemento } from "./CalculatorMemento";

export class Calculator {
    private currentValue: number = 0;

    add(value: number): void {
      this.currentValue += value;
    }

    subtract(value: number): void {
      this.currentValue -= value;
    }

    multiply(value: number): void {
      this.currentValue *= value;
    }

    divide(value: number): void {
      if (value !== 0) {
        this.currentValue /= value;
      } else {
        throw new Error("No se puede dividir por cero.");
      }
    }

    clear(): void {
      this.currentValue = 0;
    }

    getValue(): number {
      return this.currentValue;
    }

    save(): CalculatorMemento {
      return new CalculatorMemento(this.currentValue);
    }

    restore(memento: CalculatorMemento): void {
      this.currentValue = memento.getState();
    }
}